File: _unittests/ut_validation/test_cpu_fpemu.py
import unittest
from onnx_extended.ext_test_case import ExtTestCase
from onnx_extended.validation.cpu._validation import has_sse3


class TestCpuFpEmu(ExtTestCase):
    @unittest.skipIf(not has_sse3(), "SSE3 not available")
    def test_cast(self):
        from onnx_extended.validation.cpu._validation import (
            double2float_rn,
            float2half_rn,
            half2float,
        )

        self.assertEqual(double2float_rn(1), 1)
        self.assertEqual(half2float(float2half_rn(1)), 1)


if __name__ == "__main__":
    unittest.main(verbosity=2)

File: _unittests/ut_validation/test_cpu_fpemu.cpp
#include <gtest/gtest.h>
#include "onnx_extended/validation/cpu/cpu_fpemu.hpp"
#include "onnx_extended_helpers.h"
#include "onnx_extended_test_common.h"

using namespace cpu_fpemu;

TEST(validation, cast) {
#if defined(__SSSE3__)
  float f = 1.f;
  double d = 1.f;
  float ff = __double2float_rn(d);
  EXPECT_EQ(f, ff);
  unsigned short u = __float2half_rn(f);
  float bu = __half2float(u);
  EXPECT_EQ(f, bu);
#endif
}
